### DISCUSSION TITLE
Proposal: Expand supported Caller Info Attributes

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#87] [General]
Copied from https://github.com/dotnet/roslyn/issues/351

# Problem:

The currently supported caller info attributes can only provide the name of the method, the name of the source file and the line number of the call in the source file.  This is fine for scenarios such as simplifying implementations of `INotifyPropertyChanged`.  However, if using these attributes for logging the amount of information available is quite limited.
# Solution:

Expand the number of supported caller info attributes to allow embedding additional diagnostic information.  The following list is quite expansive to discuss/argue over the potential possibilities.

**CallerColumnNumberAttribute**:  The column number of where the method is invoked. 
**CallerTypeNameAttribute**:  The simple name of the declaring type of the calling method.
**CallerNamespaceAttribute**:  The namespace of the declaring type of the calling method.
**CallerFullTypeNameAttribute**:  The full name of the declaring type of the calling method.
**CallerTypeAttribute**:  The declaring type of the calling method.  This is replaced by the compiler by `ldtoken` of the type followed by a call to `Type.GetTypeFromHandle`.
**CallerMethodAttribute**:  The MethodBase of the calling method.  This is replaced by the compiler by `ldtoken` of the method reference followed by a call to `MethodBase.GetMethodFromHandle`.

Example usage:

``` cs
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;

namespace Project1 {
    public static class Program {
        private static void Foo([CallerMemberName] string memberName = null,
                                [CallerTypeName] string typeName = null,
                                [CallerNamespaceName] string namespaceName = null,
                                [CallerFullTypeName] string fullTypeName = null,
                                [CallerFilePath] string filePath = null,
                                [CallerLineNumber] int lineNumber = 0,
                                [CallerColumnNumber] int columnNumber = 0,
                                [CallerType] Type type,
                                [CallerMethod] MethodBase method)
        {
            Debug.Assert(memberName == "Main");
            Debug.Assert(typeName == "Program");
            Debug.Assert(namespaceName == "Project1");
            Debug.Assert(fullTypeName == "Project1.Program");
            Debug.Assert(filePath == "c:\\foo\\bar\\Program.cs");
            Debug.Assert(lineNumber == 29);
            Debug.Assert(columnNumber == 12);
            Debug.Assert(type == typeof(Program));
            Debug.Assert(method == typeof(Program).GetMethod("Main"));
        }

        public static void Main() {
            Foo();
        }
    }
}
```

----

It was also suggested by @AdamSpeight2008 that a caller-info struct be supported for situations where you want several of these values but don't want to have to declare/annotate multiple parameters.

```vb
Class CallerInfo
  Public ReadOnly MemberName   As String
  Public ReadOnly TypeName     As String
  Public ReadOnly Namespace    As String
  Public ReadOnly FullTypeName As String
  Public ReadOnly FilePath     As String
  Public ReadOnly LineNumber   As Integer
  Public ReadOnly ColumnNumber As Integer
  Public ReadOnly Type         As Type
  Public ReadOnly Method       As MethodBase
End Class
```

```vb
Public Sub Foo(<CallerInfo> ByVal info As CallerInfo = Nothing)
End Sub
```

LDM history:
- https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-01-03.md#other-caller-attributes

**Comment by scottdorman** [comment]
I'm not sure I see a need for `CallerColumnNumberAttribute`. What's the use case you see for that one?

By supporting just `CallerTypeAttribute`, wouldn't we gain access to the same information as what would be in `CallerTypeNameAttribute`, `CallerNamespaceAttribute`, and `CallerFullTypeNameAttribute`? 

`CallerTypeNameAttribute` is available from [`Type.Name`](https://msdn.microsoft.com/en-us/library/system.reflection.memberinfo.name(v=vs.110).aspx), `CallerNamespaceAttribute` is available from [`Type.Namespace`](https://msdn.microsoft.com/en-us/library/system.type.namespace(v=vs.110).aspx) and `CallerFullTypeNameAttribute` is available through [`Type.FullName`](https://msdn.microsoft.com/en-us/library/system.type.fullname(v=vs.110).aspx).

I do like the idea of having a `CallerInfo` struct that can be used in place of having to specify each individual attribute. This would be an "all or nothing" scenario in the sense that if I use the struct, I get all of the values filled in. If I don't want/need all of them, I can still specify them individually.

**Comment by AdamSpeight2008** [comment]
Would it be possible to make the value lazy, so they are produced when need / used.

**Comment by scottdorman** [comment]
@AdamSpeight2008 I'm not sure that would work. From [MSDN](https://msdn.microsoft.com/en-us/library/hh534540(v=vs.110).aspx), "Caller Info values are emitted as literals into the Intermediate Language (IL) at compile time." That would prevent them from being lazily evaluated since they aren't evaluated at run time.

**Comment by HaloFour** [comment]
> I'm not sure I see a need for `CallerColumnNumberAttribute`. What's the use case you see for that one?

Completes the set, mostly.  I could see some use for capturing the position of multiple calls to the same method nested in some other expression, but that might be pretty esoteric.

> By supporting just `CallerTypeAttribute`, wouldn't we gain access to the same information as what would be in `CallerTypeNameAttribute`, `CallerNamespaceAttribute`, and `CallerFullTypeNameAttribute`?

Yes.  C# already supports `CallerTypeNameAttribute`, so I built this proposal from there.

My use cases would be covered if the only caller-info attributes were `CallerMethod`, `CallerFilePath` and `CallerLineNumber`.  Everything else of value to me can be derived from that.

> I'm not sure that would work. From MSDN, "Caller Info values are emitted as literals into the Intermediate Language (IL) at compile time." That would prevent them from being lazily evaluated since they aren't evaluated at run time.

Technically true, but it's possible that some of the work could be deferred until runtime.  Specifically with `CallerInfo` where you accept a `struct` of the values.  That `struct` could have a constructor that only accepts the `RuntimeMethodHandle` for the calling method pushed by `ldtoken`, the file name and the file line number.  Then the actual method and type properties could be lazily evaluated by taking that `RuntimeMethodHandle` and loading the actual `MethodBase` and from that deriving the declaring `Type`.

**Comment by scottdorman** [comment]
> C# already supports CallerTypeNameAttribute

Do you mean `CallerMemberNameAttribute`?

So, right now we already have `CallerFilePathAttribute`, `CallerMemberNameAttribute`, and `CallerLineNumberAttribute`.

I think if we just added `CallerTypeAttribute` and `CallerMemberAttribute` (which return a `Type` and a `MemberInfo`, respectively), would that work? I changed to returning `MemberInfo` as it's the base class for `MethodBase` and I think would allow `CallerMember` to be more generally useful and work for properties as well.

> Technically true, but it's possible that some of the work could be deferred until runtime.

I would think this wouldn't be desirable due to performance considerations. It also introduces different compile-time behavior between using the attributes directly and using the struct. Then again, it might not be possible to implement `CallerTypeAttribute` and `CallerMemberAttribute` at compile time, not sure about that.

**Comment by HaloFour** [comment]
@scottdorman 

> Do you mean `CallerMemberNameAttribute`?

Oops, you're right.  None of the attributes cover the type at the moment.

> Then again, it might not be possible to implement `CallerTypeAttribute` and `CallerMemberAttribute` at compile time, not sure about that.

I don't see why not.  They're not as simple as expressions as pushing a string literal or an integer literal, but they're still relatively easy.  It's two IL opcodes to derive the `Type` or the `MethodBase` and push it onto the stack, `ldtoken` with the type or method token followed by either `call Type.GetTypeFromHandle` or `call MethodBase.GetMethodFromHandle`.  The former is literally how `typeof(T)` works.  The tokens themselves are deterministic at compile-time.

**Comment by scottdorman** [comment]
@HaloFour `typeof(T)` is a runtime call though, not a compile time literal. Are you saying that the compiler would simply emit the required IL opcodes instead of emitting the literal? I guess the difference is that by emitting the literal value, there is no runtime cost to getting the information (other than just normal parameter access overhead) but there would be slight runtime cost associated with accessing `CallerTypeAttribute` and `CallerMemberAttribute`. Granted, I doubt it's significant enough of an overhead to make much difference, especially since these probably wouldn't have a high amount of access and something could probably be done to cache the call site information, similar to what's done with `dynamic`.

**Comment by HaloFour** [comment]
@scottdorman 

Yes, it would be a runtime call.  The only way to avoid that would be to have those caller-info attributes applied to parameters of type `RuntimeTypeHandle` or `RuntimeMethodHandle` and have the compiler only emit `ldtoken` which is still evaluated at runtime but should be faster.  But all you're doing there is offloading the inevitable to the callee which is going to have to invoke those methods to get something useful anyway.

Either way, this is a *lot* less overhead than what people are doing now to support getting this kind of information which is accepting a `MethodBase` and requiring the caller to call `MethodBase.GetCurrentMethod` manually, or by getting a stack trace in the callee and trying to find the calling method, assuming that the JIT isn't inlining it.

**Comment by scottdorman** [comment]
> Either way, this is a lot less overhead than what people are doing now to support getting this kind of information

Absolutely agree on that! :) Yes, this introduces a small amount of overhead, but significantly less than what's incurred now to get the same information.

**Comment by AdamSpeight2008** [comment]
Another possibility is to do something similar to what we do with strings literals. Have a embedded dictionary lookup table.

**Comment by dotnetchris** [comment]
Building general purpose computing with CallerMemberName is extremely frustrating due to the lack of CallerTypeName, leads to all kinds of voodoo required to coerce types from somewhere that you can locate who someone is.

**Comment by dotnetchris** [comment]
@scottdorman `[CallerType]` and `[CallerMember]` would be exactly what I want.

**Comment by sextondjc** [comment]
@dotnetchris same here. Even just the fully qualified caller type name as a parameter attribute would solve a multitude of problems.

**Comment by dotnetchris** [comment]
`[CallerMember]` would be my most preferred but I would completely be able to live with only `[CallerTypeName]`

**Comment by dotnetchris** [comment]
Basically I want more and more tools to be able to write programming to write my programming.

**Comment by Pzixel** [comment]
Can't it be accomplished by #107 ? It can be 3-party nuget package that do this stuff. I'm not sure that we need built-in feature, when we just can write some extensions for Roslyn. Best issue resolution - do nothing.

**Comment by HaloFour** [comment]
@Pzixel 

Not easily.  Source generators are more about decorating entire methods and leaving their contents generally unchanged.  To implement caller attributes via source generators would require the generator to rewrite the method body so that it could replace the optional parameters.  It's certainly possible but it requires a bit more effort.  Since caller attributes are already a part of C# and work by piggy-backing off of optional parameters I think that it's worth expanding them to cover more common use cases.

**Comment by thomaslevesque** [comment]
> I changed to returning MemberInfo as it's the base class for MethodBase and I think would allow CallerMember to be more generally useful and work for properties as well.

It would be consistent with the existing `[CallerMemberName]` attribute, but it has an annoying limitation: for a property, it doesn't tell if it's the getter or setter. Maybe we should also have a `[CallerMethod]` attribute; for a method, it would do the same thing as `[CallerMember]`, but for a property it would return the getter or setter.

**Comment by 304NotModified** [comment]
Would be great if the "new version" would be backwards and forward compatible. Adding the caller member attributes is now a binary breaking change.

**Comment by yaakov-h** [comment]
@304NotModified How so?

**Comment by Happypig375** [comment]
What is the situation of this proposal? What is left to discuss? I look forward to this proposal.

**Comment by HaloFour** [comment]
@jcouv 

With `CallerArgumentExpressionAttribute` moving forward has there been any form of discussion around further expansion of the caller member attributes as outlined in this proposal?

**Comment by jcouv** [comment]
@HaloFour Good question. It may be worth consider batching some of these, if LDM is supportive.
My sense is that prioritization will be a challenge. But I'll try to get this on LDM agenda.
Added a [note](https://github.com/dotnet/roslyn/projects/27#card-6284416) to remind myself to bring this up.

**Comment by jcouv** [comment]
One more thing, LDM is largely offline until January and I expect the first couple of meetings to focus on blocking  C# 7.3 issues. So it may be mid-Jan until we can discuss this. I'll update here as I hear more.

**Comment by HaloFour** [comment]
@jcouv 

Thanks.  Note that I threw the kitchen sink in there, but this is certainly not an all-or-none proposition.

My primary use case has been within logging/tracing libraries where I'd like a cheap and easy way to figure out the context of the caller.  Currently I'm using `StackTrace` to accomplish this, which ain't cheap and is subject to being inaccurate due to JIT inlining.  And what I currently output is pretty verbose and includes the type name, generic type arguments and parameter types, formatted to be nicely readable, e.g. `Dictionary<int,string>::TryGetValue(int, out string)`

**Comment by dotnetchris** [comment]
`[CallerTypeName]` is critical, everything else is a nice to have.

**Comment by jcouv** [comment]
This was discussed in LDM today (Jan 3rd 2018). See upcoming notes.

**Comment by Pzixel** [comment]
@jcouv do you guys even sleep? It's a NY, just eat some cakes, don't meet until next monday 😋

**Comment by yaakov-h** [comment]
> See upcoming notes.

That's a bit hard at the moment. 😄

**Comment by HaloFour** [comment]
@jcouv 

https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-01-03.md#other-caller-attributes

> We are not comfortable with those just yet. We'd need to work on the details. For now we are suspicious of having a record of a lot of different information, and of the number of proposed ones in #87.

> QOTD: "What about CallerPoliticalAffiliation?".

Understood.  To note, I threw a lot of spaghetti at the wall in this proposal.  I don't intend for all of it to stick.

If you wanted to effectively capture all of the use cases that could be covered here, `CallerMethodAttribute` I believe would give you access to all of the other pieces of information via reflection, e.g. method name, simple type name, full type name, namespace, assembly, etc.

**Comment by TheXenocide** [comment]
At least having access to the full type name would definitely have helped with several diagnostic utilities I've made. I can see the merit of using Type/MemberInfo and would have used them were they available, though given the discussion about runtime evaluation versus compile-time literals, there may be some merit to having string attributes. Do we know why some values were provided with the original feature release where others weren't? I strongly suspect this was included due to increasing reliance on INotifyPropertyChanged with no fully-baked native AOP mechanism (which is something I'd find extremely appealing and could similarly solve this and many other problems abstractly--with a more generally useful range of functionality--though it's likely much easier to implement an analog to compile-time functionality that already exists).

**Comment by dotnetchris** [comment]
@TheXenocide https://github.com/dotnet/csharplang/issues/164 upon completion will give us absolute AOP at every single point in dotnet since you'll be able to proxy any object transparently to the existing receivers.

**Comment by TheXenocide** [comment]
@dotnetchris, I'll have to give that a more thorough read when I have more time, but at first glance while I can see how it solves many types of problems that AOP frameworks tend to target (INotifyPropertyChange, multiple-inheritance stand-ins, type-specific wrappers, etc.), it does not appear to universally solve compile-time instruction enhancement in all the ways and with all the optimizations that existing aspect weavers do (and, especially notably, does not provide compile-time evaluation of programmatic logic to apply aspects; e.g. PostSharp's IAspectProvider). I'm not sure how those shapes/extensions would be used to, say, allow a developer to convert an annotation (like an Attribute applied to any parameter on any method) into a set of conditional logic based on information provided by the caller at compile-time the way these Caller Info Attributes work (I could implement this with PostSharp or Fody today, with some caveats that would be better solved using ASTs/syntax analysis à la Roslyn). I look forward to additional language improvements to help solve cross-cutting concerns, but there are still fundamental AOP concepts not addressed from what I can tell.

**Comment by dotnetchris** [comment]
@TheXenocide right, it's not compile time interception it would be run time interception. 

>  allow a developer to convert an annotation (like an Attribute applied to any parameter on any method)

you would do something like setup a static constructor or use one of the App_Start files etc etc to bootstrap that would do assembly scanning, find all the methods with that attribute. You'd construct some type of dictionary or map of that data. Then at run time you flow your code over the shape and you would do things if the flow hit your map.

Then you would likely combine this with Dependency Injection that you're invisibly injecting your Shape proxies to wherever you want them and they appear to be the traditional type to the receiving sites.

**Comment by TheXenocide** [comment]
Alas, I'm much more interested in compile time aspects than runtime proxies, though I understand how these might be useful.

**Comment by dotnetchris** [comment]
Any news on this?

**Comment by KyleKolpek** [comment]
I would also love an update on this. Particularly `CallerTypeName`.

Also, for my use case it would be important to have the type an Attribute is attributed to, rather than the true caller. This is consistent with [`CallerMemberName`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callermembernameattribute?view=netframework-4.7.2#remarks) behavior (see quote below). I expect this would be the case, but also wanted to make sure it was noted.

> Attribute constructor


> The name of the member to which the attribute is applied. If the attribute is any element within a member (such as a parameter, a return value, or a generic type parameter), this result is the name of the member that's associated with that element.

**Comment by alexsorokoletov** [comment]
+1 for this proposal. CallerTypeName is the thing (as many other pointed), other are less important. Current workaround (getting filename and then deduct type name from the file) is a bit messy.

**Reply by RazvanRotaru** [reply]
the issue with this is that you may have a `partial class` that is not named properly. it requires you to pay unnecessary attention.

**Comment by maxfridbe** [comment]
Any news?

**Comment by jcouv** [comment]
> Any news?

Our hands are full at the moment with current C# 8 feature set (in particular, nullable reference types and default interface members).

**Comment by abezverkov** [comment]
+1 for CallerType or CallerTypeName.   Just came searching again for the ump-teenth time in case something had changed.

**Comment by georgiuk** [comment]
What about CallerInstanceObject so as to pass the instance of the object that called the Method/Property. In case of static class this will be always null.

**Comment by yaakov-h** [comment]
That sounds like a fun can of worms.

**Comment by TheXenocide** [comment]
@georgiuk  Not sure how I feel about CallerInstanceObject; that seems like logic that would generally be more explicit (and not sure how useful it would be as it would probably have to be of type "Object"). The other attributes are strictly metadata, whereas this crosses into the actual data territory. Feels like extension methods are analogous and more appropriate representation.

**Comment by georgiuk** [comment]
Yes, CallerInstanceObject probably is off the limits for this API.   
Another thing I was thinking is if there will be possible to make it more strict. I mean right now a caller can bypass it and pass anything else as parameter. Wouldn't be useful if the compiler disallow passing those parameters explicitly?
Having CallerMethod to discriminate Getter and Setter would also be useful.

**Comment by CodeAngry** [comment]
By 2025 maybe we get this. Because when you design this CallerInfo you add...

- CallerFilePath
- CallerLineNumber
- CallerMemberName

...but somehow manage to not even think about the CallerTypeName that has the CallerMemberName, because... who needs that?!

**Another one of those Microsoft moments in API design.** That will be eventually fixed _(to be read: made to work as it should have had since start)_ in a decade.

Who knows, maybe someone comes up with the idea of changing CallerMemberName from `Method` to `ParentNS[.ChildNS].Type.Method`. Insane creativity right here. And then every logging code will actually work properly.

**Comment by TheXenocide** [comment]
@CodeAngry - username checks out...

**Comment by TonyValenti** [comment]
I wanted to post my note for this, especially "CallerMemberType" as a Type (and not a string).

**Comment by tstrickler4** [comment]
This might be a whole other can of worms, but I think it would be extremely helpful if CallerMemberType could be used in generics as well. For example, you could have some function like this:

    T MyFunc<[CallerMemberType] T>()
    {
        //do stuff
    }

Then you could use it like so:

    public MyClass MyProperty
    {
        get => MyFunc();
    }

**Comment by taooceros** [comment]
Any information about CallerTypeName in .Net 6?

**Reply by CodeAngry** [reply]
This is a 2017 request. So I'd say check back in 2027,

**Comment by cremor** [comment]
Related: #4984

**Comment by frankhaugen** [comment]
This really need some traction, as I keep coming back to needing this functionality when dealing with things like logging, or I'm working on "plugin-esque" applications.

I keep ending up creating messy "hacks" to get the information I need to meet some specs, and if this proposal was implemented I probalbly would have saved tens, if not hundreds of hours of development and maintainence of convoluted reflection -based logic

**Reply by CyrusNajmabadi** [reply]
FWIW, this post isn't super helpful.  I cannot glean from it what in particular you need, and verify that adding it will be sufficient for you to "meet some specs".  

It would be great to give crisp examples of what you're trying to do, how you do it currently, and which attributes in particular would make a substantive difference for you.

Thanks!

**Reply by dotnetchris** [reply]
We need `[CallerMemberTypeName]` that's the bare minimum. 

Optimally we get `[CallerMember]` and `[CallerMemberType]`

**Reply by CyrusNajmabadi** [reply]
what would those return?

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

As stated [earlier](https://github.com/dotnet/csharplang/discussions/87#discussioncomment-64949) in this discussion:

> My use cases would be covered if the only caller-info attributes were CallerMethod, CallerFilePath and CallerLineNumber. Everything else of value to me can be derived from that.

Two are already covered, the missing one would be `CallerMethod`.  That could return a `MethodBase` or a `RuntimeMethodHandle`.  I believe every other value above could be derived from that, except for CallerColumnNumber, but that one might not be of value.

**Reply by cremor** [reply]
@CyrusNajmabadi I aggree with @dotnetchris that `[CallerMemberTypeName]` is the bare minimum requirement, but only if the following two points are true:
- It returns the `FullName` of the type.
- It never returns the name of a compiler generated type.

If any of those points can't be fulfilled we need `[CallerMember]` (provides a `MethodBase` or `RuntimeMethodHandle`) or `[CallerMemberType]` (provides a `Type`) to be able to get the "real" full name ourselves.

**Comment by CodeAngry** [comment]
So, will 2022 be the year?!

**Reply by frankhaugen** [reply]
Maybe if we do it ourselves and raise a PR with awesome docs, tests and performance numbers. Do you have the free-time and skills for it?

**Comment by dotnetchris** [comment]
> Maybe if we do it ourselves and raise a PR with awesome docs, tests and performance numbers. Do you have the free-time and skills for it?

I feel like multiple sources have been submitted for this over the past half decade

**Comment by perahoky** [comment]
we are interested in CallerTypeNameAttribute too. Would be very nice for logging performancetracing etc 
thank you

**Comment by digitaldias** [comment]
We are also interested in the CallerTypeName attribute, specifically for logging needs. Would be nice to know what the current state of this is

**Reply by TonyValenti** [reply]
Related to call type name, it would be nice if that was smart enough to return either this.GetType or 
Typeof(TSelf) (for static classes)

**Reply by CyrusNajmabadi** [reply]
> Would be nice to know what the current state of this is

Nothing has changed since hte last LDM meeting on the topic.  If anything changes we will update this discussion.  :)

**Comment by njsokalski** [comment]
I like the idea of a CallerInfo class, it would make debugging universally easier because it is not restricted to one specific aspect like the other Caller attributes

**Comment by AntonKrug** [comment]
CallerTypeName or CallerType would make things so much easier

**Comment by WahidBitar** [comment]
Lets keep this up.
`CallerTypeName`

**Reply by TonyValenti** [reply]
I would like just a plain old caller type.

**Reply by TheXenocide** [reply]
Agreed, type would be more useful, you can get the name from the type easily enough (and decide whether you want fully qualified or short, include info about generic type params or not, etc.)

**Comment by xparadoxical** [comment]
Turns out I too need `[CallerTypeName]` (context: [a helper used in async tests](https://github.com/xparadoxical/gd-apis/blob/b116dc46eb9b58a2c6602cd41886831afa9fb6b6/GeometryDash.Tests/Server/Serialization/Generator/GeneratorTestHelper.cs#L32-L38)). Parsing it out from `[CallerFilePath]` should be good enough™️ in my case, but isn't guaranteed to be correct, since you can have >1 type in a file. In async methods it's impossible to get the correct StackFrame from a [StackTrace](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.stacktrace). So my only options are to rely on CallerFilePath, specify filenames manually or have unpleasant `nameof`s sprinkled around.

**Comment by ahdung** [comment]
only `CallerTypeName` needed for me, any news plz?

**Comment by codemonkey493** [comment]
@CodeAngry 
> By 2025 maybe we get this

Nope, we don't have it in 2025 yet. Apparently, the C# team is too busy implementing features that nobody needs.

**Reply by colejohnson66** [reply]
While I, personally, am disappointed many of my desired features won't be implemented this time around, I am also very excited for extensions.

**Reply by codemonkey493** [reply]
@colejohnson66 I've yet to see anyone who needed `buffer = { [^1] = 0 }`

**Reply by dmitry-azaraev** [reply]
> @colejohnson66 I've yet to see anyone who needed `buffer = { [^1] = 0 }`

It sounds like irony. I did not like this fancy indexing stuff, however, CallerTypeName add no value. All attributes like CallerFilePath - they optional. Your program MUST work as expected with them or without them, and this is useful only for disgnostic purposes. So it is absolutely clear what feature which have zero use-cases are not progressed. More over: this must not be implemented, all showcases of such feat is stupidly fucked code, instead of doing things like they should.

**Reply by CyrusNajmabadi** [reply]
Please keep this discussion on topic.  Discussions about other features, and your particular desire around them aren't appropriate to the discussion of the design here.  Thanks.

**Comment by DomenPigeon** [comment]
Sometimes there are some special classes which need to do some initialization and no other class should do that, for those cases `CallerMemberType` would really be useful, as we have the `nameof` attribute already (for which refactoring works nicely) it could really be as simple as a string:

```csharp
public class InitializedTypeGetter
{
    private InitializedType? _init;

    public void Init(InitializedType init, [CallerMemberType] string? caller = null)
    {
        if (_init != null)
        {
            throw new InvalidOperationException("Already initialized");
        }

        if (caller != nameof(MySpecialInitializator))
        {
            throw new InvalidOperationException($"Can only be initialized by {nameof(MySpecialInitializator)}");
        }

        _init = init;
    }

    public InitializedType Init=> _init ?? throw new InvalidOperationException("Not yet initialized");
}
```

**Reply by colejohnson66** [reply]
That feels… wrong. A type shouldn’t care who’s using it. If you want initialization to only occur from one place, don’t expose that initialization logic.

**Reply by DomenPigeon** [reply]
I don't aggree, there is a way to do exactly this already, you just need to put the `InitializedTypeGetter` and the `MySpecialInitializator` together in a new project (assembly) and then make the `Init` function `internal` and you get the desired functinality, which is perfectly acceptable. But that just isn't always practical and also having too many projects extends the build time.

**Reply by colejohnson66** [reply]
Again: an “Init” function shouldn’t care who calls it. If the parameters are ok, then they’re ok.

**Reply by cyraid** [reply]
@DomenPigeon Couldn't you just use Generics? Something like:

```csharp
public void Init<T>() {
  _init = typeof(T); // Could also use typeof(T).Name etc.
  if (_init != typeof(MySpecialInitializator) && !_init.IsSubclassOf(typeof(MySpecialInitializator)))
    throw new InvalidOperationException($"Can only be initialized by {nameof(MySpecialInitializator)}");
} // Procedure //
```

You could always use Generic constraints too:

```csharp
public void Init<T>() where T : MySpecialInitializator {
} // Procedure //
```

**Comment by jnm2** [comment]
For those wanting CallerTypeName: would https://github.com/dotnet/csharplang/discussions/9017 suit your use case? You would write `[CallerIdentity] Type callerType = null`, and get a System.Type instance off of which you could do `.FullName` to get the metadata name, e.g. ``SomeNamepace.OuterType`1+InnerType``. Or you could just access `.Name`, or `.Namespace`.

**Reply by HaloFour** [reply]
Would scratch the itch for me.  Even better if `CallerIdentity` could also be applied to a parameter of type `MethodBase`/`MethodInfo`.

**Reply by jnm2** [reply]
That proposal would allow it for `Assembly`, `Type`, and `MethodBase` parameters.
